File: Testing/tests/client/Tests/BasicCSharp/TestVirtualMethods.cs
using Bridge;
using Bridge.QUnit;

namespace ClientTestLibrary
{
    class TestVirtualMethods
    {
        [FileName("testVirtualMethods.js")]
        class A
        {
            public virtual string Test()
            {
                return "A";
            }
        }

        [FileName("testVirtualMethods.js")]
        class B : A
        {
            public string TestA()
            {
                return base.Test();
            }

            public override string Test()
            {
                return "B";
            }
        }
        public static void TestB(Assert assert)
        {
            assert.Expect(7);

            var a = new A();

            assert.Ok(a != null, "Instance of A created");
            assert.Equal(a.Test(), "A", "a.Test() = 'A'");


            var b = new B();

            assert.Ok(b != null, "Instance of B created");
            assert.Equal(b.Test(), "B", "b.Test() = 'B'");
            assert.Equal(b.TestA(), "A", "b.TestA() = 'A'");

            A c = new B();

            assert.Ok(c != null, "Instance of C created");
            assert.Equal(c.Test(), "B", "c.Test() = 'B'");
        }
    }
}

File: Testing/www/js/testVirtualMethods.js
/* global Bridge */

"use strict";

Bridge.define('ClientTestLibrary.TestVirtualMethods', {
    statics: {
        testB: function (assert) {
            assert.expect(7);

            var a = new ClientTestLibrary.TestVirtualMethods.A();

            assert.ok(a !== null, "Instance of A created");
            assert.equal(a.test(), "A", "a.Test() = 'A'");

            var b = new ClientTestLibrary.TestVirtualMethods.B();

            assert.ok(b !== null, "Instance of B created");
            assert.equal(b.test(), "B", "b.Test() = 'B'");
            assert.equal(b.testA(), "A", "b.TestA() = 'A'");

            var c = new ClientTestLibrary.TestVirtualMethods.B();

            assert.ok(c !== null, "Instance of C created");
            assert.equal(c.test(), "B", "c.Test() = 'B'");
        }
    }
});

Bridge.define('ClientTestLibrary.TestVirtualMethods.A', {
    test: function () {
        return "A";
    }
});

Bridge.define('ClientTestLibrary.TestVirtualMethods.B', {
    inherits: [ClientTestLibrary.TestVirtualMethods.A],
    testA: function () {
        return ClientTestLibrary.TestVirtualMethods.A.prototype.test.call(this);
    },
    test: function () {
        return "B";
    }
});



Bridge.init();